### DISCUSSION TITLE
How to block a certain domain from using my script

### DISCUSSION THREAD

**Discussion by 0PandaDEV** [plausible/analytics#4307] [Self-Hosted Support] [answered]
Hello I have someone that copied my website 1:1 which is in itself absolutely disgusting but this person is so dumb he even forgot to remove my plausible analytics script so now I want to block his domain from modifying my analytics with his website.

**Comment by ruslandoga** [comment, marked as answer]
👋 @0PandaDEV 

I think you can do it with the "Shields" feature. It should be available at `/:site/settings/shields/hostnames`

For example, for me it's at https://who.edify.space/stats.edify.space/settings/shields/hostnames

**Reply by ruslandoga** [reply]
More info: https://plausible.io/docs/excluding#exclude-visits-by-hostname

**Reply by 0PandaDEV** [reply]
I don't have the shields tab in my analytics page
![image](https://github.com/plausible/analytics/assets/70103896/34dcd071-26dd-4a56-9260-ccc26e215864)

**Reply by ruslandoga** [reply]
Then you need to upgrade to a recent version, the most recent right now is [v2.1.1.](https://github.com/plausible/analytics/discussions/4195)

Note that depending on the version you are currently using you might need to perform multiple upgrades, e.g. v2.0.0 -> v2.1.0 -> v2.1.1.

**Reply by 0PandaDEV** [reply]
how do i do this

**Reply by 0PandaDEV** [reply]
Also for some reason I don't have the community edition?
![image](https://github.com/plausible/analytics/assets/70103896/9f5e5e15-ec8e-4aa2-83ab-0b76fb7a5801)

**Reply by ruslandoga** [reply]
New releases use ghcr.io, old releases used docker.io.

You will need to upgrade using the following guides:
- v2.0 -> v2.1.0: https://github.com/plausible/analytics/discussions/4125
- v2.1.0 -> v2.1.1: https://github.com/plausible/analytics/discussions/4195

**Reply by 0PandaDEV** [reply]
ok

**Reply by 0PandaDEV** [reply]
I tried upgrading, and it won't start back up when I go and hook into the logs of the hosting-plausible-1 container, I get this: 

```
dockerERROR! Config provider Config.Reader failed with:
** (ArgumentError) incorrect padding
    (elixir 1.16.0) lib/base.ex:735: Base.decode64base!/2
    /app/releases/0.0.1/runtime.exs:163: (file)
    (elixir 1.16.0) src/elixir.erl:405: :elixir.eval_external_handler/3
    (stdlib 5.2) erl_eval.erl:750: :erl_eval.do_apply/7
    (stdlib 5.2) erl_eval.erl:1026: :erl_eval.expr_list/7
    (stdlib 5.2) erl_eval.erl:456: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:539: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:325: :erl_eval.expr/6

Runtime terminating during boot ({#{message=><<"incorrect padding">>,'__struct__'=>'Elixir.ArgumentError','__exception__'=>true},[{'Elixir.Base','decode64base!',2,[{file,"lib/base.ex"},{line,735}]},{elixir_eval,'__FILE__',1,[{file,"/app/releases/0.0.1/runtime.exs"},{line,163}]},{elixir,eval_external_handler,3,[{file,"src/elixir.erl"},{line,405}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,750}]},{erl_eval,expr_list,7,[{file,"erl_eval.erl"},{line,1026}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,456}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,539}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,325}]}]})

Crash dump is being written to: erl_crash.dump...ERROR! Config provider Config.Reader failed with:
** (ArgumentError) incorrect padding
    (elixir 1.16.0) lib/base.ex:735: Base.decode64base!/2
    /app/releases/0.0.1/runtime.exs:163: (file)
    (elixir 1.16.0) src/elixir.erl:405: :elixir.eval_external_handler/3
    (stdlib 5.2) erl_eval.erl:750: :erl_eval.do_apply/7
    (stdlib 5.2) erl_eval.erl:1026: :erl_eval.expr_list/7
    (stdlib 5.2) erl_eval.erl:456: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:539: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:325: :erl_eval.expr/6

Runtime terminating during boot ({#{message=><<"incorrect padding">>,'__struct__'=>'Elixir.ArgumentError','__exception__'=>true},[{'Elixir.Base','decode64base!',2,[{file,"lib/base.ex"},{line,735}]},{elixir_eval,'__FILE__',1,[{file,"/app/releases/0.0.1/runtime.exs"},{line,163}]},{elixir,eval_external_handler,3,[{file,"src/elixir.erl"},{line,405}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,750}]},{erl_eval,expr_list,7,[{file,"erl_eval.erl"},{line,1026}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,456}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,539}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,325}]}]})

Crash dump is being written to: erl_crash.dump...ERROR! Config provider Config.Reader failed with:
** (ArgumentError) incorrect padding
    (elixir 1.16.0) lib/base.ex:735: Base.decode64base!/2
    /app/releases/0.0.1/runtime.exs:163: (file)
    (elixir 1.16.0) src/elixir.erl:405: :elixir.eval_external_handler/3
    (stdlib 5.2) erl_eval.erl:750: :erl_eval.do_apply/7
    (stdlib 5.2) erl_eval.erl:1026: :erl_eval.expr_list/7
    (stdlib 5.2) erl_eval.erl:456: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:539: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:325: :erl_eval.expr/6

Runtime terminating during boot ({#{message=><<"incorrect padding">>,'__struct__'=>'Elixir.ArgumentError','__exception__'=>true},[{'Elixir.Base','decode64base!',2,[{file,"lib/base.ex"},{line,735}]},{elixir_eval,'__FILE__',1,[{file,"/app/releases/0.0.1/runtime.exs"},{line,163}]},{elixir,eval_external_handler,3,[{file,"src/elixir.erl"},{line,405}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,750}]},{erl_eval,expr_list,7,[{file,"erl_eval.erl"},{line,1026}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,456}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,539}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,325}]}]})

Crash dump is being written to: erl_crash.dump...ERROR! Config provider Config.Reader failed with:
** (ArgumentError) incorrect padding
    (elixir 1.16.0) lib/base.ex:735: Base.decode64base!/2
    /app/releases/0.0.1/runtime.exs:163: (file)
    (elixir 1.16.0) src/elixir.erl:405: :elixir.eval_external_handler/3
    (stdlib 5.2) erl_eval.erl:750: :erl_eval.do_apply/7
    (stdlib 5.2) erl_eval.erl:1026: :erl_eval.expr_list/7
    (stdlib 5.2) erl_eval.erl:456: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:539: :erl_eval.expr/6
    (stdlib 5.2) erl_eval.erl:325: :erl_eval.expr/6

Runtime terminating during boot ({#{message=><<"incorrect padding">>,'__struct__'=>'Elixir.ArgumentError','__exception__'=>true},[{'Elixir.Base','decode64base!',2,[{file,"lib/base.ex"},{line,735}]},{elixir_eval,'__FILE__',1,[{file,"/app/releases/0.0.1/runtime.exs"},{line,163}]},{elixir,eval_external_handler,3,[{file,"src/elixir.erl"},{line,405}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,750}]},{erl_eval,expr_list,7,[{file,"erl_eval.erl"},{line,1026}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,456}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,539}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,325}]}]})

Crash dump is being written to: erl_crash.dump...ER
```

**Reply by ruslandoga** [reply]
Please check that your TOTP_VAULT_KEY is correct. It's explained how to generate it in https://github.com/plausible/analytics/discussions/4125

For reference, the app fails here: https://github.com/plausible/analytics/blob/379714c133f832400eb2777425f083281f80a4aa/config/runtime.exs#L163-L165

**Reply by 0PandaDEV** [reply]
I did generate one and set it in the plausible-conf.env, but it had a different format than the one in the docs

it starts with opensslz XXXXXX

**Reply by ruslandoga** [reply]
Please post your OpenSSL version and the key and the command you used here.

**Reply by 0PandaDEV** [reply]
VERSION: OpenSSL 3.2.1 30 Jan 2024 (Library: OpenSSL 3.2.1 30 Jan 2024)
COMMAND: `openssl rand -base64 32`
KEY: `opensslz8XxnqRW17vUgf47la1lQESUbCyPDD7opT4SQx8naIE=`

**Reply by ruslandoga** [reply]
`z8XxnqRW17vUgf47la1lQESUbCyPDD7opT4SQx8naIE=` is a valid TOTP_VAULT_KEY, 32 bytes encoded as base64.

You'll need to check your OpenSSL installation later, it's not supposed to add `openssl` prefix to random data.

**Reply by 0PandaDEV** [reply]
Fixed it by generating the key with my Mac which has the newest version it not completely works again and is up

**Reply by 0PandaDEV** [reply]
Ok but now back to my original problem I tried adding a hostname whitelist like this `*.pandadev.net` so that only scripts that are placed on websites from pandadev.net are allowed to use the analytics but for some reason it still allows dev.quadratik.pro

**Reply by ruslandoga** [reply]
Right. The request that `dev.quadratik.pro` makes - when "copied as cURL" from the browser - is

```
curl 'https://plausible.pandadev.net/api/event' \
-X 'POST' \
-H 'Content-Type: text/plain' \
-H 'Accept: */*' \
-H 'Sec-Fetch-Site: cross-site' \
-H 'Accept-Language: en-US,en;q=0.9' \
-H 'Accept-Encoding: gzip, deflate, br' \
-H 'Sec-Fetch-Mode: cors' \
-H 'Host: plausible.pandadev.net' \
-H 'Origin: https://dev.quadratik.pro' \
-H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15' \
-H 'Referer: https://dev.quadratik.pro/' \
-H 'Content-Length: 77' \
-H 'Connection: keep-alive' \
-H 'Sec-Fetch-Dest: empty' \
--data-binary '{"n":"pageview","u":"https://dev.quadratik.pro/","d":"pandadev.net","r":null}'
```

So unless the Hostnames Shield feature also blocks by origins or referrers or "u" or "d" from the payload, it wouldn't work since the hostname is `plausible.pandadev.net` which is allowed.

**Reply by ruslandoga** [reply]
Seems like it uses "u" (short for "url") for the checks, so it should work.

https://github.com/plausible/analytics/blob/379714c133f832400eb2777425f083281f80a4aa/lib/plausible/ingestion/request.ex#L182-L192

https://github.com/plausible/analytics/blob/379714c133f832400eb2777425f083281f80a4aa/lib/plausible/ingestion/request.ex#L194-L202

https://github.com/plausible/analytics/blob/379714c133f832400eb2777425f083281f80a4aa/lib/plausible/ingestion/event.ex#L202-L208

https://github.com/plausible/analytics/blob/379714c133f832400eb2777425f083281f80a4aa/lib/plausible/shields.ex#L32-L45
